File: server/src/main/java/com/tomato/bookstore/service/NoteService.java
package com.tomato.bookstore.service;

import com.tomato.bookstore.constant.FeedbackType;
import com.tomato.bookstore.dto.NoteCommentCreateDTO;
import com.tomato.bookstore.dto.NoteCommentDTO;
import com.tomato.bookstore.dto.NoteCreateDTO;
import com.tomato.bookstore.dto.NoteDTO;
import java.util.List;

/** 读书笔记服务接口 */
public interface NoteService {
  /**
   * 根据商品 ID 获取所有笔记
   *
   * @param productId 商品 ID
   * @return 笔记列表
   */
  List<NoteDTO> getNotesByProductId(Long productId);

  /**
   * 根据用户 ID 获取所有笔记
   *
   * @param userId 用户 ID
   * @return 笔记列表
   */
  List<NoteDTO> getNotesByUserId(Long userId);

  /**
   * 获取所有笔记（管理员使用）
   *
   * @return 笔记列表
   */
  List<NoteDTO> getAllNotes();

  /**
   * 根据笔记 ID 获取笔记详情
   *
   * @param noteId 笔记 ID
   * @param currentUserId 当前用户 ID（可选，提供后可获取用户反馈状态）
   * @return 笔记详情
   */
  NoteDTO getNoteById(Long noteId, Long currentUserId);

  /**
   * 创建笔记
   *
   * @param productId 商品 ID
   * @param userId 用户 ID
   * @param noteCreateDTO 笔记创建 DTO
   * @return 创建的笔记
   */
  NoteDTO createNote(Long productId, Long userId, NoteCreateDTO noteCreateDTO);

  /**
   * 更新笔记
   *
   * @param noteId 笔记 ID
   * @param userId 用户 ID
   * @param noteCreateDTO 笔记创建 DTO
   * @return 更新后的笔记
   */
  NoteDTO updateNote(Long noteId, Long userId, NoteCreateDTO noteCreateDTO);

  /**
   * 管理员更新笔记
   *
   * @param noteId 笔记 ID
   * @param noteCreateDTO 笔记创建 DTO
   * @return 更新后的笔记
   */
  NoteDTO updateNoteByAdmin(Long noteId, NoteCreateDTO noteCreateDTO);

  /**
   * 删除笔记
   *
   * @param noteId 笔记 ID
   * @param userId 用户 ID
   */
  void deleteNote(Long noteId, Long userId);

  /**
   * 管理员删除笔记
   *
   * @param noteId 笔记 ID
   */
  void deleteNoteByAdmin(Long noteId);

  /**
   * 为笔记添加或更新反馈
   *
   * @param noteId 笔记 ID
   * @param userId 用户 ID
   * @param feedbackType 反馈类型
   * @return 更新后的笔记
   */
  NoteDTO addOrUpdateFeedback(Long noteId, Long userId, FeedbackType feedbackType);

  /**
   * 获取笔记的所有评论
   *
   * @param noteId 笔记 ID
   * @return 评论列表
   */
  List<NoteCommentDTO> getNoteComments(Long noteId);

  /**
   * 添加笔记评论
   *
   * @param noteId 笔记 ID
   * @param userId 用户 ID
   * @param commentCreateDTO 评论创建 DTO
   * @return 创建的评论
   */
  NoteCommentDTO addComment(Long noteId, Long userId, NoteCommentCreateDTO commentCreateDTO);

  /**
   * 删除笔记评论
   *
   * @param commentId 评论 ID
   * @param userId 用户 ID
   */
  void deleteComment(Long commentId, Long userId);

  /**
   * 管理员删除笔记评论
   *
   * @param commentId 评论 ID
   */
  void deleteCommentByAdmin(Long commentId);
}

File: client/src/services/note-service.ts
import { NOTE_MODULE } from "@/constants/apiPrefix";
import type { ApiResponse } from "@/types/api";
import type {
  Note,
  NoteComment,
  NoteCommentCreateParams,
  NoteCreateParams,
  NoteFeedbackParams,
  NoteUpdateParams,
} from "@/types/note";
import apiClient from "@/utils/apiClient";

/**
 * 读书笔记服务
 */
const noteService = {
  /**
   * 获取指定商品的所有读书笔记
   *
   * @param {number} productId 商品 ID
   * @returns {Promise<ApiResponse<Note[]>>} 笔记列表
   */
  getProductNotes(productId: number): Promise<ApiResponse<Note[]>> {
    return apiClient.get(`${NOTE_MODULE}/product/${productId}`);
  },

  /**
   * 获取当前用户的所有读书笔记
   *
   * @returns {Promise<ApiResponse<Note[]>>} 笔记列表
   */
  getUserNotes(): Promise<ApiResponse<Note[]>> {
    return apiClient.get(`${NOTE_MODULE}/user`);
  },

  /**
   * 获取指定用户的所有读书笔记（仅管理员）
   *
   * @param {number} userId 用户 ID
   * @returns {Promise<ApiResponse<Note[]>>} 笔记列表
   */
  getUserNotesByAdmin(userId: number): Promise<ApiResponse<Note[]>> {
    return apiClient.get(`${NOTE_MODULE}/user/${userId}`);
  },

  /**
   * 获取所有读书笔记（仅管理员）
   *
   * @returns {Promise<ApiResponse<Note[]>>} 笔记列表
   */
  getAllNotes(): Promise<ApiResponse<Note[]>> {
    return apiClient.get(`${NOTE_MODULE}/all`);
  },

  /**
   * 获取读书笔记详情
   *
   * @param {number} noteId 笔记 ID
   * @returns {Promise<ApiResponse<Note>>} 笔记详情
   */
  getNoteById(noteId: number): Promise<ApiResponse<Note>> {
    return apiClient.get(`${NOTE_MODULE}/${noteId}`);
  },

  /**
   * 创建读书笔记
   *
   * @param {number} productId 商品 ID
   * @param {NoteCreateParams} params 笔记参数
   * @returns {Promise<ApiResponse<Note>>} 创建的笔记
   */
  createNote(
    productId: number,
    params: NoteCreateParams,
  ): Promise<ApiResponse<Note>> {
    return apiClient.post(`${NOTE_MODULE}/product/${productId}`, params);
  },

  /**
   * 更新读书笔记
   *
   * @param {number} noteId 笔记 ID
   * @param {NoteUpdateParams} params 更新参数
   * @returns {Promise<ApiResponse<Note>>} 更新后的笔记
   */
  updateNote(
    noteId: number,
    params: NoteUpdateParams,
  ): Promise<ApiResponse<Note>> {
    return apiClient.put(`${NOTE_MODULE}/${noteId}`, params);
  },

  /**
   * 管理员更新读书笔记
   *
   * @param {number} noteId 笔记 ID
   * @param {NoteUpdateParams} params 更新参数
   * @returns {Promise<ApiResponse<Note>>} 更新后的笔记
   */
  updateNoteByAdmin(
    noteId: number,
    params: NoteUpdateParams,
  ): Promise<ApiResponse<Note>> {
    return apiClient.put(`${NOTE_MODULE}/admin/${noteId}`, params);
  },

  /**
   * 删除读书笔记
   *
   * @param {number} noteId 笔记 ID
   * @returns {Promise<ApiResponse<string>>} 删除结果
   */
  deleteNote(noteId: number): Promise<ApiResponse<string>> {
    return apiClient.delete(`${NOTE_MODULE}/${noteId}`);
  },

  /**
   * 管理员删除读书笔记
   *
   * @param {number} noteId 笔记 ID
   * @returns {Promise<ApiResponse<string>>} 删除结果
   */
  deleteNoteByAdmin(noteId: number): Promise<ApiResponse<string>> {
    return apiClient.delete(`${NOTE_MODULE}/admin/${noteId}`);
  },

  /**
   * 为读书笔记添加反馈（点赞/点踩）
   *
   * @param {number} noteId 笔记 ID
   * @param {NoteFeedbackParams} params 反馈参数
   * @returns {Promise<ApiResponse<Note>>} 更新后的笔记
   */
  addFeedback(
    noteId: number,
    params: NoteFeedbackParams,
  ): Promise<ApiResponse<Note>> {
    return apiClient.post(`${NOTE_MODULE}/${noteId}/feedback`, params);
  },

  /**
   * 获取读书笔记的所有评论
   *
   * @param {number} noteId 笔记 ID
   * @returns {Promise<ApiResponse<NoteComment[]>>} 评论列表
   */
  getNoteComments(noteId: number): Promise<ApiResponse<NoteComment[]>> {
    return apiClient.get(`${NOTE_MODULE}/${noteId}/comments`);
  },

  /**
   * 为读书笔记添加评论
   *
   * @param {number} noteId 笔记 ID
   * @param {NoteCommentCreateParams} params 评论参数
   * @returns {Promise<ApiResponse<NoteComment>>} 创建的评论
   */
  addComment(
    noteId: number,
    params: NoteCommentCreateParams,
  ): Promise<ApiResponse<NoteComment>> {
    return apiClient.post(`${NOTE_MODULE}/${noteId}/comments`, params);
  },

  /**
   * 删除读书笔记评论
   *
   * @param {number} noteId 笔记 ID
   * @param {number} commentId 评论 ID
   * @returns {Promise<ApiResponse<string>>} 删除结果
   */
  deleteComment(
    noteId: number,
    commentId: number,
  ): Promise<ApiResponse<string>> {
    return apiClient.delete(`${NOTE_MODULE}/${noteId}/comments/${commentId}`);
  },

  /**
   * 管理员删除读书笔记评论
   *
   * @param {number} noteId 笔记 ID
   * @param {number} commentId 评论 ID
   * @returns {Promise<ApiResponse<string>>} 删除结果
   */
  deleteCommentByAdmin(
    noteId: number,
    commentId: number,
  ): Promise<ApiResponse<string>> {
    return apiClient.delete(
      `${NOTE_MODULE}/admin/${noteId}/comments/${commentId}`,
    );
  },
};

export default noteService;
